### PR TITLE
Add dateName function

### PR DESCRIPTION
**Pull request by dankondr**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added dateName function.

Detailed description / Documentation draft:

Added dateName function which returns specified date part of a Date, DateTime or DateTime64 object as a string.

This PR implements #7325

### CODE REVIEW

**Comment by dankondr** [comment]
@alexey-milovidov

File: tests/queries/0_stateless/01811_datename.sql [hunk 1]
```diff
@@ -0,0 +1,59 @@
+SELECT dateName('year', toDateTime('2021-04-14 11:22:33'));
```
Comment (kitaisreal) [inline tests/queries/0_stateless/01811_datename.sql:1]: Generate functional test reference file.
Check https://clickhouse.tech/docs/en/development/tests/#adding-a-new-test

File: src/Functions/dateName.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 3)
+            throw Exception(
```
Comment (kitaisreal) [inline src/Functions/dateName.cpp:60]: Consider to use fmt like Exception throw.
```
throw Exception(
     ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, 
    "Number of arguments for function {} doesn't match: passed {}", 
     getName(), 
     toString(arguments.size())).
```

Comment (dankondr) [reply src/Functions/dateName.cpp:62]: Fixed

File: tests/queries/0_stateless/01811_datename.sql (hunk 1 above)
Comment (dankondr) [reply tests/queries/0_stateless/01811_datename.sql:1]: Added

**Comment by dankondr** [comment]
Is integration and stress test failures relate to my changes? @kitaisreal

**Comment by kitaisreal** [comment]
@dankondr, no there is only Style Check error that need to be fixed.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Almost LGTM need to address some minor issues.

File: src/Functions/dateName.cpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+private:
+    template <typename Time>
+    class DatePartWriter
+    {
+    public:
+        void writeDatePart(char *& target, const String & datepart, Time source, const DateLUTImpl & timezone)
+        {
+            datepart_functions.at(datepart)(target, source, timezone);
```
[the remark this answers is not in the document]
Comment (dankondr) [reply src/Functions/dateName.cpp:180]: I made it to escape numerous strcmp. How can I replicate it using enum?

File: src/Functions/dateName.cpp [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+        static inline void writeQuarter(char *& target, Time source, const DateLUTImpl & timezone)
+        {
+            writeNumber(target, ToQuarterImpl::execute(source, timezone));
+        }
+
+        static inline void writeMonth(char *& target, Time source, const DateLUTImpl & timezone)
+        {
+            const auto month = ToMonthImpl::execute(source, timezone);
+            const String monthnames[12]
```
[the remark this answers is not in the document]
Comment (dankondr) [reply src/Functions/dateName.cpp:212]: How is it declared?

File: src/Functions/dateName.cpp [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        const ColumnConst * datepart_column = checkAndGetColumnConst<ColumnString>(arguments[0].column.get());
+        if (!datepart_column)
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first ('datepart') argument of function {}. Must be constant string.",
+                arguments[0].column->getName(),
+                getName());
+
+        using T = typename ActionValueTypeMap<DataType>::ActionValueType;
+        auto datepart_writer = DatePartWriter<T>();
```
[the remark this answers is not in the document]
Comment (dankondr) [reply src/Functions/dateName.cpp:119]: As there is no problem in using implementations for DateTime64 we should use Int64, so ActionValueTypeMap is the most readable solution

File: src/Functions/dateName.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (danlark1) [reply src/Functions/dateName.cpp:212]: `static constexpr std::string_view[] = {"January", ...}` or `static constexpr std::array = {"January"sv, ...`}

**Pull request merged**